db/db_compaction_test.cc

```diff
@@ -10623,6 +10623,45 @@ TEST_F(DBCompactionTest, ReleaseCompactionDuringManifestWrite) {
   SyncPoint::GetInstance()->ClearAllCallBacks();
 }
 
+TEST_F(DBCompactionTest, RecordNewestKeyTime) {
```

Comment:
One motivation for adding newest key time is that TTL compaction won't be accurate after a file went through temperature change compaction. Specifically, the creation_time table property will change from "newest key time" to "oldest_key_time" and a file can qualify for TTL compaction too early. Is it possible to write a unit test that tests this scenario?

Reply:
I updated this test case (now called `RecordNewestKeyTimeForTtlCompaction`) to cover TTL logic. Let me know what you think. It still validates that we are populating `newest_key_time` on flush, and that the max of the input file values is being used during compaction. Now this test case also validates that the TTL will respect the `newest_key_time` and that it does not drop files that would have otherwise have been dropped using the old `creation_time` / `oldest_ancestor_time` field

Reply:
>  it does not drop files that would have otherwise have been dropped using the old creation_time / oldest_ancestor_time field

The `creation_time` table property will be the same as newest key time unless a file goes through a compaction though. I was thinking of the following scenario: 

Current time T.
File 1: creation time T, oldest_ancestor_time T-4.
File 2: creation time T-10, oldest_ancestor_time T-15.

Assume TTL is 12, then no file qualifies for TTL yet. If File 2 is compacted for temperature change, its `creation_time` will be set to T-15 as the oldest_ancestor_time. It now incorrectly qualifies for TTL compaction.

Reply:
Hmm @cbi42 could you check this test case again? This test case does compact files, so the `newest_key_time` gets updated based on the input files. 

I also updated the test case to make it very clear that the `creation_time` is equal to the `oldest_ancestor_time`, which is less than the `newest_key_time` after compaction. 

So I think this should cover what you want, unless you specifically want to use temperature change compaction for this test?

Reply:
(see everything in the test case after this block)
```cpp
  // After compaction, the newest_key_time of the output file should be the max
  // of the input files
```

Reply:
Ok, I see that an intra-L0 compaction for all 4 L0 files happens after reopen. I think temperature change compaction is more explicit and is used by some users compared to allow_compaction. But this is fine. 

Btw, I think you'll need to set `SetTimeElapseOnlySleepOnReopen()` before reopen, otherwise the test will be flaky (time still elapses in addition to the mock sleep).

Reply:
> Ok, I see that an intra-L0 compaction for all 4 L0 files happens after reopen. I think temperature change compaction is more explicit and is used by some users compared to allow_compaction. But this is fine.
👍 

> Btw, I think you'll need to set SetTimeElapseOnlySleepOnReopen() before reopen, otherwise the test will be flaky (time still elapses in addition to the mock sleep).
Ah, good to know. Updated the PR.